pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/DataTableBlock.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.runtime.blocks;
+
+import java.io.IOException;
+import org.apache.pinot.common.utils.DataTable;
+import org.apache.pinot.core.common.Block;
+import org.apache.pinot.core.common.BlockDocIdSet;
+import org.apache.pinot.core.common.BlockDocIdValueSet;
+import org.apache.pinot.core.common.BlockMetadata;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.operator.blocks.InstanceResponseBlock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A {@code DataTableBlock} is a row-based data block backed by a {@link DataTable}.
+ */
+public class DataTableBlock implements Block {
```

Comment:
We should discuss this a little bit more before going down the row oriented path

I think we will end up rewriting all the existing columnar operators into their corresponding row-wise counterparts in this new multi-stage query engine and that is probably a lot of unnecessary / avoidable work if we could just use a columnar / record batch like wire format for exchange messages between executors for shuffle / partition / broadcast and at least try to use existing columnar operators. 

I think it is ok to see if reusing existing columnar operators is possible in the new engine or may be just copy them over here and do minimal changes as needed. So it's probably ok to have 2 different versions of columnar operators (if reuse is absolutely impossible) -- one for existing engine and one for new engine

What is more concerning to me is that since we are columnar (storage format, inmemory format and execution wise), I think we should stick with that for performance (and related benefits in OLAP) reasons because otherwise we will find ourselves doing another rewrite

- rewrite all current columnar operators into rowwise for multi-engine to work
- rewrite all multi-engine row-wise operators at a later point in future into columnar counterparts to improve performance (I am certain this will be one of the things targeted later to improve performance)

Reply:
good idea. let's put a TODO (or design discussion section on this)

Reply:
we should probably estimate whether it is better to send data over the wire using columnar format
1. there's tradeoff when encoding data in row vs. columnar format when key-partitioned block split operation is done (for example distributed hash-join)
2. there's tradeoff when data arrived at the receiving end and being processed (obviously columnar format wins in many of the situation)
in this case let's come up with a columnar format as well as a row-based format (for the data block sent over mailbox)

because if columnar format turns up performing better, we don't need to reinvent a row-base operator code

Reply:
This is another TODO item